File: src/math/polygon.ts
import Line from "./line";
import LineSegment from "./line-segment";
import { compare, max, min } from "./number";
import Rectangle from "./rectangle";
import Vector2 from "./vector2";

class Polygon {
  public readonly points: readonly Vector2[];

  private _boundingRectangle?: Rectangle;
  private _sides?: readonly LineSegment[];

  public get boundingRectangle(): Rectangle {
    if (this._boundingRectangle === undefined) {
      this._boundingRectangle = this.computeBoundingRectangle();
    }

    return this._boundingRectangle;
  }

  public get sides(): readonly LineSegment[] {
    if (this._sides === undefined) {
      this._sides = this.computeSides();
    }

    return this._sides;
  }

  public constructor(points: readonly Vector2[]) {
    this.points = points;
  }

  public containsPoint(point: Vector2): boolean {
    if (!this.boundingRectangle.containsPoint(point)) return false;

    const line = new Line(0, 1, -point.y); // Line parallel to X-axis

    let intersectionsCount = 0;

    for (let i = 0; i < this.sides.length; i++) {
      const side = this.sides[i];

      if (side.b.equals(point)) {
        // We will get intersection with next side
        continue;
      }

      const intersectionPoint = line.getIntersection(side.line);

      if (
        intersectionPoint &&
        compare(intersectionPoint.x, point.x) >= 0 &&
        side.boundingRectangle.containsPoint(intersectionPoint)
      ) {
        intersectionsCount++;
      }
    }

    return intersectionsCount % 2 === 1;
  }

  private computeSides(): readonly LineSegment[] {
    const result = [];

    for (let i = 1; i < this.points.length; i++) {
      result.push(new LineSegment(this.points[i - 1], this.points[i]));
    }

    result.push(
      new LineSegment(this.points[this.points.length - 1], this.points[0])
    );

    return result;
  }

  private computeBoundingRectangle(): Rectangle {
    const startPoint = this.points[0];

    let minX = startPoint.x;
    let minY = startPoint.y;
    let maxX = startPoint.x;
    let maxY = startPoint.y;

    for (let i = 1; i < this.points.length; i++) {
      const point = this.points[i];

      minX = min(minX, point.x);
      maxX = max(maxX, point.x);
      minY = min(minY, point.y);
      maxY = max(maxY, point.y);
    }

    return new Rectangle(minX, minY, maxX - minX, maxY - minY);
  }
}

export default Polygon;

File: logic/src/math/polygon.rs
use once_cell::unsync::OnceCell;

use super::{compare, Line, LineSegment, Rectangle, Vector2};

pub struct Polygon {
    points: Vec<Vector2<f32>>,
    bounding_rectangle: OnceCell<Rectangle<f32>>,
    sides: OnceCell<Vec<LineSegment>>,
}

impl Polygon {
    pub fn new(points: Vec<Vector2<f32>>) -> Self {
        if points.len() < 3 {
            panic!("At least 3 points expected.");
        }

        Self {
            points,
            bounding_rectangle: OnceCell::new(),
            sides: OnceCell::new(),
        }
    }

    pub fn points(&self) -> &[Vector2<f32>] {
        &self.points
    }

    pub fn bounding_rectangle(&self) -> &Rectangle<f32> {
        if self.bounding_rectangle.get().is_none() {
            let start_point = self.points[0];

            let mut min_x = start_point.x;
            let mut min_y = start_point.y;
            let mut max_x = start_point.x;
            let mut max_y = start_point.y;

            for point in &self.points[1..] {
                min_x = f32::min(min_x, point.x);
                max_x = f32::max(max_x, point.x);
                min_y = f32::min(min_y, point.y);
                max_y = f32::max(max_y, point.y);
            }

            _ = self.bounding_rectangle.set(Rectangle::new(
                min_x,
                min_y,
                max_x - min_x,
                max_y - min_y,
            ));
        }

        self.bounding_rectangle.get().unwrap()
    }

    pub fn sides(&self) -> &[LineSegment] {
        if self.sides.get().is_none() {
            let mut sides = Vec::with_capacity(self.points.len());

            for i in 1..self.points.len() {
                sides.push(LineSegment::new(self.points[i - 1], self.points[i]));
            }

            sides.push(LineSegment::new(
                *self.points.last().unwrap(),
                *self.points.first().unwrap(),
            ));

            _ = self.sides.set(sides);
        }

        self.sides.get().unwrap()
    }

    pub fn contains_point(&self, point: &Vector2<f32>) -> bool {
        if !self.bounding_rectangle().contains_point(point) {
            return false;
        }

        let line = Line {
            a: 0.0,
            b: 1.0,
            c: -point.y,
        }; // Line parallel to X-axis

        let mut intersections_count = 0;

        for side in self.sides() {
            let intersection_point = line.intersection(side.line());

            if let Some(intersection_point) = intersection_point {
                if compare(intersection_point.x, point.x).is_ge()
                    && side
                        .bounding_rectangle()
                        .contains_point(&intersection_point)
                {
                    if intersection_point.equals(point) {
                        return true;
                    }

                    intersections_count += 1;
                }
            }
        }

        intersections_count % 2 == 1
    }
}
